Add `search_name_sku_or_global_unique_id` parameter to products REST API

staskus opened the pull request:
### Changes proposed in this Pull Request:

RFC: pfZP8i-cI-p2 -worth checking it first
Related changes: https://github.com/woocommerce/woocommerce/pull/57494
WOOMOB-713

This pull request builds on the `search_name_or_sku` parameter by adding a new `search_name_sku_or_global_unique_id` argument to the REST API products endpoint (`GET /wc/v3/products`). This extends the existing tokenized search functionality to include Global Unique ID (GTIN, UPC, EAN, or ISBN) alongside product names and SKUs.

The search behavior is consistent with the existing `search_name_or_sku` parameter:
*   Tokenized: The argument is treated as a collection of space-separated tokens.
*   Same-Field Matching: All tokens must be present within the same field for a product to match. A product matches if all tokens are found in the name OR all tokens are found in the SKU OR all tokens are found in the Global Unique ID.
*   Partial Match: A word containing a token is a valid match.

**Examples:** 

- An argument value of `blue shirt` will match a product with the name "Casual Blue Shirt" because both tokens exist in the same field (name), but not "Red Shirt". 
- `scarf 987` will NOT match a product with name "Premium Wool Scarf" and GTIN "9876543210" because the tokens exist in different fields.

This new argument takes precedence over all other search-related parameters: `search_name_or_sku`, `search`, `sku`, and `search_sku`. If `search_name_sku_or_global_unique_id` is present in a request, all others will be ignored.

This functionality is added to the v3 REST API products endpoint only.

### How to test the changes in this Pull Request:

1.  Create products with the following details:
    *   Product A:
        *   Name: `Premium Wool Scarf`
        *   SKU: `SCARF-W-PREM`
        *   GTIN: `9876543210`
    *   Product B:
        *   Name: `Casual Blue Shirt`
        *   SKU: `SHIRT-B-CASUAL`
        *   GTIN: `1234567890`
    *   Product C:
        *   Name: `Red Cotton Tee`
        *   SKU: `TEE-C-RED`
        *   GTIN: `0000000000`

2.  Create an application password for your WordPress user in your user profile page from the WP admin area.

3.  **Test same-field matching:** Perform the following request to find "Product B" by searching for tokens that exist in the same field (name).

    ```bash
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=blue+shirt"
    ```

    You should get one result: the "Casual Blue Shirt" product.

4.  **Test individual field searches:** Verify that single tokens work across all fields.

    ```bash
    # Find by name token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=scarf"

    # Find by SKU token  
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=SCARF"

    # Find by GTIN token
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=987"
    ```

5.  **Verify same-field requirement:** Confirm that cross-field searches return no results (this is expected behavior).

    ```bash
    # This should return 0 results (tokens in different fields)
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=scarf+987"
    ```

6.  **Verify precedence:** Confirm that `search_name_sku_or_global_unique_id` overrides all other search parameters.

    ```bash
    # Overrides search_name_or_sku
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=blue+shirt&search_name_or_sku=scarf"

    # Overrides search and search_sku
    curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku,global_unique_id&search_name_sku_or_global_unique_id=blue+shirt&search=red&search_sku=TEE"
    ```

7.  **Verify regression:** Confirm that the older search parameters continue to work as expected when the new one is not present.

    *   The following returns the "Casual Blue Shirt" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search_name_or_sku=blue+shirt"
        ```

    *   The following returns the "Red Cotton Tee" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search=red"
        ```

    *   The following returns the "Premium Wool Scarf" product:
        ```bash
        curl -u <user>:<application_password> "https://localhost/wp-json/wc/v3/products?_fields=name,sku&search_sku=SCARF"
        ```


### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add `search_name_sku_or_global_unique_id` parameter to the Products REST API endpoint.

</details>

staskus committed:
Add search_name_sku_or_gtin parameter to products REST API

Implements a new search parameter for the WooCommerce products REST API endpoint
that searches across product names, SKUs, and GTIN (Global Trade Item Number) fields.

Key features:
- Tokenized search with AND logic between tokens, OR logic between fields
- Takes precedence over search, sku, search_sku, and search_name_or_sku parameters
- Uses existing wc_product_meta_lookup table for performance optimization
- Supports partial matching across all three field types
- Handles edge cases like disabled SKUs

Implementation includes test coverage for basic functionality, parameter precedence,
and edge cases. Follows existing WooCommerce coding standards and patterns.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

staskus committed:
Refactor product search API parameters and improve search logic

- Renamed search_name_sku_or_gtin parameter to search_name_sku_or_unique_id for consistency with global_unique_id field naming
- Added improved string validation for search parameters (is_string and trim checks)
- Refactored search query building logic into dedicated build_search_clauses method for better maintainability
- Updated all related variable names, comments, and documentation to use global_unique_id terminology
- Updated PHPUnit tests to reflect the new parameter name and maintain test coverage
- Improved code organization by extracting complex search logic into a reusable private method

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

staskus committed:
Rename search parameter for API consistency

- Renamed search_name_sku_or_unique_id to search_name_sku_or_global_unique_id
- Ensures perfect consistency with existing global_unique_id parameter
- Removes ambiguity between unique_id and global_unique_id terminology
- Updated all related variable names, comments, and documentation
- Updated PHPUnit tests to reflect the new parameter name
- Maintains backward compatibility while improving API clarity

staskus committed:
Merge branch 'pr-59241' into woomob-713-add-global_unique_id-to-searched-fields

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -311,9 +319,20 @@ protected function prepare_objects_query( $request ) {
+		$search_name_or_sku_arg                  = $request['search_name_or_sku'] ?? '';
 
-		if ( '' !== $search_name_or_sku_arg ) {
+		if ( is_string( $search_name_sku_or_global_unique_id_arg ) && '' !== trim( $search_name_sku_or_global_unique_id_arg ) ) {
```
Given the REST arg is configured to be a `string` I don't think it can happen that this isn't a string. If you pass an array, for example, validation for the request will fail. Also, we could apply the `trim()` to `$search_name_sku_or_global_unique_id_arg` directly and simplify the check here to just `'' !== $search_name_sku_or_global_unique_id_arg` or even just `if ( $search_name_sku_or_global_unique_id_arg )`.

staskus replied:
Got it, thanks for the explanation!

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -472,19 +493,35 @@ protected function get_objects( $query_args ) {
+		} elseif ( ! empty( $this->search_sku_arg_value ) ) {
+			// For SKU search, we need the meta lookup table.
+			$needs_join = true;
+		}
```
Instead of all the various branches maybe we can rewrite this along the lines of

```php
$needs_join =
	$this->search_name_sku_or_global_unique_id_tokens
	|| ( $this->search_name_or_sku_tokens && wc_product_sku_enabled() )
	|| ( ! empty( $this->search_sku_arg_value ) );
```

staskus replied:
True, much simpler! 👍

jorgeatorres commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php:
```diff
@@ -496,40 +533,71 @@ public function add_search_criteria_to_wp_query_join( $join ) {
+	 * @param bool  $include_unique_id Whether to include global_unique_id search.
+	 * @return string
+	 */
+	private function build_search_clauses( $tokens, $include_sku, $include_unique_id ) {
```
I'm not sure the clauses being built here match the PR description. For example, when the search term is `scarf 987` the final query contains this:

```sql
WHERE [...]
(
	((wp_posts.post_title LIKE '%scarf%') AND (wp_posts.post_title LIKE '%987%'))
	OR ((wc_product_meta_lookup.sku LIKE '%scarf%') AND (wc_product_meta_lookup.sku LIKE '%987%'))
	OR ((wc_product_meta_lookup.global_unique_id LIKE '%scarf%') AND (wc_product_meta_lookup.global_unique_id LIKE '%987%'))
)
[...]
```

This wouldn't return a match for the test case `scarf 987` because neither field (title, SKU, or GTIN) contain _both_ terms `scarf` and `987`. My understanding is that it should return a result if any of the fields match at least one of the terms.

Am I missing something?

staskus replied:
> This wouldn't return a match for the test case scarf 987 because neither field (title, SKU, or GTIN) contain both terms scarf and 987

This is the right observation. I made a mistake with the testing description. The functionality should work in a way that was already developed in https://github.com/woocommerce/woocommerce/pull/57494 We don't want to match terms across different fields

The code works as we want it to work:

- `blue shirt` will match a product with the name "Casual Blue Shirt" because both tokens exist in the same field (name), but not "Red Shirt".
- `scarf 987` will NOT match a product with name "Premium Wool Scarf" and GTIN "9876543210" because the tokens exist in different fields.

jorgeatorres reviewed (changes requested):
Hey @staskus!

Thanks for the PR. I've left some comments in the code, but my main concern is that this doesn't seem to be working as described in the testing instructions but it might just be a situation where we need to flip some ANDs and ORs 😸 .

Please let me know what you think.

staskus committed:
Refactor join logic to use cleaner conditional structure

Replace verbose if-else branches with concise boolean expression for determining
when to join wc_product_meta_lookup table based on search criteria.

staskus committed:
Simplify search parameter validation logic

Remove unnecessary is_string() check and apply trim() directly to the parameter.
The REST API validation ensures the parameter is a string, so the type check
is redundant.

staskus commented:
@jorgeatorres, I appreciate the review!

I got confused myself with multiple search terms and multiple teams, but the intention is to keep it as it is. We want all terms to be present within the same field for a product to match.

Also thanks for the PHP simplifications. I work mostly with Swift, so I don't immediately see those opportunities.

Based on pfZP8i-cI-p2, I will separately look into a more flexible approach while trying to keep as much inner functionality that has been built here as possible. That's why I'm not re-requesting a review yet.

Thanks again!

staskus committed:
Merge branch 'trunk' into woomob-713-add-global_unique_id-to-searched-fields
